src/persist-client/src/lib.rs

```diff
@@ -539,6 +554,187 @@ impl PersistClient {
         Ok(writer)
     }
 
+    /// Returns a [BatchBuilder] that can be used to write a batch of updates to
+    /// blob storage which can then be appended to the given shard using
+    /// [Self::compare_and_append_batch] or [Self::append_batch], or which can
+    /// be read using [PersistClient::read_batches_consolidated].
+    ///
+    /// The builder uses a bounded amount of memory, even when the number of
+    /// updates is very large. Individual records, however, should be small
+    /// enough that we can reasonably chunk them up: O(KB) is definitely fine,
+    /// O(MB) come talk to us.
+    #[instrument(level = "debug", fields(shard = %shard_id))]
+    pub async fn batch_builder<K, V, T, D>(
```

Comment:
These are basically duplicated code. I could add a `WriteHandle::batch_builder_inner` that doesn't take a `self` but all the arguments it needs. And then both `WriteHandle` and the code here could call that one.

Same for `read_batches_consolidated` below.

I didn't do this for now because I felt that the method signature was almost the largest part of these methods. But happy to change that!